src/mir/construction.md

```diff
@@ -1 +1,111 @@
 # MIR construction
+
+The lowering of [HIR] to [MIR] occurs for the following (probably incomplete)
+list of items:
+
+* Function and Closure bodies
+* Initializers of `static` and `const` items
+* Initializers of enum discriminants
+* Glue and Shims of any kind
+    * Tuple struct initializer functions
+    * Drop code (the `Drop::drop` function is not called directly)
+    * Drop implementations of types without an explicit `Drop` implementation
+
+The lowering is triggered by calling the `mir_built` query. The entire lowering
+code lives in `src/librustc_mir/build`. There is an intermediate representation
+between [HIR] and [MIR] called the `HAIR` that is only used during the lowering.
+The `HAIR` has datatypes that mirror the [HIR] datatypes, but instead of e.g. `-x`
+being a `hair::ExprKind::Neg(hair::Expr)` it is a `hair::ExprKind::Neg(hir::Expr)`.
+
+This shallowness enables the `HAIR` to represent all datatypes that [HIR] has, but
+without having to create an in-memory copy of the entire [HIR]. The `HAIR` also
+does a few simplifications, e.g. method calls and function calls have been merged
+into a single variant.
+
+The lowering creates local variables for every argument as specified in the signature.
+Next it creates local variables for every binding specified (e.g. `(a, b): (i32, String)`)
+produces 3 bindings, one for the argument, and two for the bindings. Next it generates
+field accesses that read the fields from the argument and writes the value to the binding
+variable.
+
+With this initialization out of the way, the lowering triggers a recursive call
+to a function that generates the MIR for the body (a `Block` expression) and
+writes the result into the `RETURN_PLACE`.
+
+## `unpack!` all the things
+
+One important thing of note is the `unpack!` macro, which accompanies all recursive
+calls. The macro ensures, that you get the result of the recursive call while updating
+the basic block that you are now in. As an example: lowering `a + b` will need to do
+three somewhat independent things:
+
+* give you an `Rvalue` referring to the result of the operation
+* insert an assertion ensuring that the operation does not overflow
+* tell you in which basic block you should write further operations into, because
+  the basic block has changed due to the inserted assertion (assertions terminate
+  blocks and jump either to a panic block or a newly created block, the latter being
+  the one you get back).
+
+The `unpack!` macro will call the recursive function you pass it, return the `Rvalue`
+and update the basic block by mutating the basic block variable you pass to it.
+
+## Lowering expressions into the desired MIR
+
+There are essentially four kinds of representations one might want of a value:
```

Comment:
I would be careful with the phrase "value" -- I think we should be moving towards a terminology of expressions appearing in a "place context" (where they name some memory location) and "value context" (where we extracting the value). Hence I would maybe say, "Depending on the context in which an expression appears, we may want to lower it into one of four different sorts of things:" or something.